### PR TITLE
Koski oid management dialog

### LINKED ISSUE #673: Koski: Management dialog
Create management dialog which
* Makes it easier to manage all the Koski-OIDs
* Allows invalidation of study permits that have failed to transfer correctly

### PR DESCRIPTION
**Pull request by anttiviljakainen**
Invalidation is done so that the whole student is loaded from Koski, the other study permits are removed but the invalidated one, which is marked for invalidation and the whole thing is sent back to the server.

The major change needed was the ability to read the JSON from Koski to the Java model and most of the changes are related to that change. Polymorphism of the API data is pretty big problem here which results in the Deserializer classes which try to figure out which class to instantiate.

In nutshell
* Superclasses (usually abstract) have JsonDeserialize(using = xxDeserializer) -annotation to define the deserializer for the class tree
* If superclass has deserializer, it needs to be overwritten in subclasses with JsonDeserialize(using = None) or you get stack overflow
* JsonInclude(NON_NULL) tells serializer (pojo -> json) to ignore (not write) properties which are null
* JsonIgnoreProperties(ignoreUnknown = true) tells deserializer (json -> pojo) to ignore properties that are in json but not in pojo (without this throws exception)
* Enums used for KoodistoViite had to be listed in a utility class and they needed reverseLookup -method if the enums had values
* Model classes need non-parameterized constructors and thus no uninitialized final fields

Also removed 'tila'(/suorituksentila)-field from several places as it has been deprecated from the API.

Closes #673

### CODE REVIEW

File: pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/InvalidateKoskiStudyPermitJSONRequestController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    Person person = personDAO.findById(personId);
+    String oid = requestContext.getString("oid");
+    
+    KoskiClient client = CDI.current().select(KoskiClient.class).get();
+    
+    try {
+      client.invalidateStudyOid(person, oid);
+    } catch (Exception e) {
+      // TODO Auto-generated catch block
+      e.printStackTrace();
```
Comment (anttiviljakainen) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/InvalidateKoskiStudyPermitJSONRequestController.java]: Maybe something here

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+public class KoskiConsts {
+
+  public static class VariableNames {
+    public static final String KOSKI_HENKILO_OID = "koski.henkilo-oid";
+    public static final String KOSKI_STUDYPERMISSION_ID = "koski.studypermission-id";
+    public static final String KOSKI_LINKED_STUDYPERMISSION_ID = "koski.linked-to-studypermission-id";
+  }
+  
+  public static String getStudentIdentifier(KoskiStudyProgrammeHandler handler, Long studentId) {
+    return handler.name() + ":" + String.valueOf(studentId);
```
Comment (pkukkon) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java]: return KoskiConsts.getStudentIdentifier(handler, studentId); perhaps?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/model/deserializers/KoodistoViiteDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    
+    return null;
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  private Enum reverseLookup(Class enumClass, String enumValue) {
+    try {
+      Method reverseLookupMethod = enumClass.getMethod("reverseLookup", String.class);
+
+      if (reverseLookupMethod != null) {
```
Comment (pkukkon) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/model/deserializers/KoodistoViiteDeserializer.java]: I suppose null check is irrelevant, as getMethod would have thrown an exception about a missing method already?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOppiaineenArviointi.java [hunk 4]
```diff
@@ -9,6 +9,9 @@
 
 public class LukionOppiaineenArviointi {
 
+  public LukionOppiaineenArviointi() {
```
Comment (pkukkon) [inline framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOppiaineenArviointi.java]: This one doesn't need JsonDeserialize(using = JsonDeserializer.None.class)?

File: pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/EditKoskiPersonVariablesJSONRequestController.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+public class EditKoskiPersonVariablesJSONRequestController extends JSONRequestController {
+
+  public void process(JSONRequestContext requestContext) {
+    StudentDAO studentDAO = DAOFactory.getInstance().getStudentDAO();
+    PersonDAO personDAO = DAOFactory.getInstance().getPersonDAO();
+    UserVariableDAO userVariableDAO = DAOFactory.getInstance().getUserVariableDAO();
+    PersonVariableDAO personVariableDAO = DAOFactory.getInstance().getPersonVariableDAO();
+
+    Long personId = NumberUtils.createLong(requestContext.getRequest().getParameter("personId"));
```
Comment (pkukkon) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/json/students/EditKoskiPersonVariablesJSONRequestController.java]: Wouldn't requestContext.getLong("personId") do the trick?

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java (hunk 2 above)
Comment (anttiviljakainen) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java]: ? :)

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOppiaineenArviointi.java (hunk 4 above)
Comment (anttiviljakainen) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/model/lukio/LukionOppiaineenArviointi.java]: Nope, it's only for subclasses where parent class has a custom deserializer.

File: framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java (hunk 2 above)
Comment (pkukkon) [reply framework/src/main/java/fi/otavanopisto/pyramus/koski/KoskiConsts.java]: Ahem ^^;

**Pull request merged**